File: src/main/java/core/procedures/characters/CharPredicate.java
package core.procedures.characters;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;

import java.util.function.Predicate;

public final class CharPredicate extends AFn {

  public static final CharPredicate IS_CHAR_WHITESPACE = new CharPredicate("char-whitespace?", Character::isWhitespace);
  public static final CharPredicate IS_CHAR_ALPHABETIC = new CharPredicate("char-alphabetic?", Character::isAlphabetic);
  public static final CharPredicate IS_CHAR_UPPER_CASE = new CharPredicate("char-upper-case?", Character::isUpperCase);
  public static final CharPredicate IS_CHAR_LOWER_CASE = new CharPredicate("char-lower-case?", Character::isLowerCase);
  public static final CharPredicate IS_CHAR_NUMERIC    = new CharPredicate("char-numeric?",    Character::isDigit);

  private final String name;
  private final Predicate<Character> predicate;

  private CharPredicate(String name, Predicate<Character> predicate) {
    super(new FnArgsBuilder().minArgs(1).maxArgs(1).mandatoryArgsTypes(new Class[]{Character.class}));
    this.name = name;
    this.predicate = predicate;
  }

  @Override
  public Boolean apply1(Object arg) {
    return predicate.test((Character)arg);
  }

  @Override
  public String getName() {
    return name;
  }
}

File: src/main/java/core/procedures/characters/CharPredicate.kt
package core.procedures.characters

import core.procedures.AFn
import core.procedures.Arity.Exactly

class CharPredicate private constructor(override val name: String, private inline val predicate: (Char) -> Boolean) :
        AFn<Char?, Boolean>(arity = Exactly(1), mandatoryArgsTypes = arrayOf(Char::class.javaObjectType)) {

    companion object {
        val IS_CHAR_WHITESPACE  = CharPredicate("char-whitespace?",  Character::isWhitespace)
        val IS_CHAR_ALPHABETIC  = CharPredicate("char-alphabetic?",  { Character.isAlphabetic(it.toInt()) })
        val IS_CHAR_UPPER_CASE  = CharPredicate("char-upper-case?",  Character::isUpperCase)
        val IS_CHAR_LOWER_CASE  = CharPredicate("char-lower-case?",  Character::isLowerCase)
        val IS_CHAR_NUMERIC     = CharPredicate("char-numeric?",     Character::isDigit)
        val IS_CHAR_TITLE_CASE  = CharPredicate("char-title-case?",  Character::isTitleCase)
        val IS_CHAR_ISO_CONTROL = CharPredicate("char-iso-control?", Character::isISOControl)
    }

    override operator fun invoke(arg: Char?) = predicate(arg!!)
}
